scipy/special/special/cephes/ndtr.h

```diff
@@ -259,12 +259,16 @@ namespace cephes {
         x = a * M_SQRT1_2;
         z = std::abs(x);
 
-        if (z < M_SQRT1_2) {
+        if (z < 1.0) {
             y = 0.5 + 0.5 * erf(x);
         } else {
             y = 0.5 * erfc(z);
             if (x > 0) {
-                y = 1.0 - y;
+                t = 1.0 - y;
+		if (t == 1.0 && y != 0) {
+		    set_error("ndtr", SF_ERROR_LOSS, NULL);
+		}
+		return t;
```

Comment:
The nature of floating point numbers is that they are not evenly spaced. They fan out logarithmically in such a way that they are very dense near zero, and thin out towards $\pm \infty$.

I think an example will help clear things up. Take for instance `z = 20`, then `y = 0.5 * erfc(z)` will be `2.6979328058039506e-176`. But `t = `1.0 - y` will be 1.0, because floating point numbers are not dense enough near 1.0 to represent a number so close to 1.0.

Please remove this change. Also, in general, one typically should not expect exact equality to hold when working with floating point numbers just because two expressions are mathematically identical (e.g. it's possible for `x + (y + z)` to be different from `(x + y) + z`, so one typically avoids exact comparisons. 

If you're interested in learning more, I'd suggest checking out the wikipedia pages for [Machine Epsilon](https://en.wikipedia.org/wiki/Machine_epsilon) and [unit in the last place](https://en.wikipedia.org/wiki/Unit_in_the_last_place). [What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html) by David Goldberg is another great resource.i

Reply:
Thank @steppi, I know. But :

1. It is not an issue that ```ndtr(1e-18) == 0.5```, even though it does ```ndtr(1e-18)``` -> ```0.5 + erf(1e-18)``` -> ```0.5 + 1.1283791670955125e-18``` -> ```0.5``` Indeed, ```sc.ndtri(sc.ndtr(1e-18)) == 0.0``` which is ok.

2. However, it **IS** an issue that ```ndtr(25) == 1.0```. The difference between the two examples is that the codomain of a CDF is (0,1), not [0,1].  Indeed, ```sc.ndtri(sc.ndtr(25)) == Inf``` which is **NOT** ok.

ndtr(1e-18) is imprecise because of floating-point arithmetic but 0.5 falls within the codomain of a cdf, so it is not an issue. But ndtr(25) raises a problem beyond simple accuracy of floating-point arithmetic : it gives values outside of the expected codomain of the function.

Actually, **the current version of erf and of erfc DO raise a warning in similar situations** and start warning at around abs(x) == 27 when the result is outside the codomain. This change simply extends this to ndtr. You can test :

```
import pytest
import scipy.special as sc
        
with sc.errstate(all='warn'):
    with pytest.warns(sc.SpecialFunctionWarning):
        sc.erfc(27)
        
with sc.errstate(all='warn'):
    with pytest.warns(sc.SpecialFunctionWarning):
        sc.erfc(-27)

with sc.errstate(all='warn'):
    with pytest.warns(sc.SpecialFunctionWarning):
        sc.erf(27)
        
with sc.errstate(all='warn'):
    with pytest.warns(sc.SpecialFunctionWarning):
        sc.erf(-27)
```

Reply:
I understand what you're saying, but from a floating point perspective, I think 1.0 is the correct value for `ndtr(25)` since it is the closest `double` to the actual real number answer. `1.0` isn't in the image of `ndtr` when the domain is the real numbers, but it is when the domain is the set of double precision IEEE-754 floating point numbers, which we can think of as their own algebraic structure. The `ndtr` in SciPy is an approximation to "ideal" correctly rounded `ndtr` on IEEE-754, and I think the `SF_ERROR_LOSS` is intended for situations where the relative error between the correctly rounded result and the computed result is undesirably large.

I don't think of `1.0` occupying a privileged place in floating point arithmetic compared to any other value which is not infinite, zero, or `nan`, so if we need to warn for loss of precision even when producing a correctly rounded result here, it seems like it would also be necessary to produce such a warning for any floating point computation which produces an inexact result, even if correctly rounded. The warnings that occur for `erf` and `erfc` are due to intermediate underflow in calculations, not due to an inexact result being produced.

Reply:
```suggestion
                y = 1.0 - y;
```

The relative error when the result is 1.0 instead of some number slightly less than 1.0 (which doesn't exist in `float64`) is less than machine precision. The code `SF_ERROR_LOSS` means "loss of precision"

https://github.com/scipy/scipy/blob/3226c32bb268777b97652bd07ed2d2cc284b6b5a/scipy/special/xsf/error.h#L3-L15

So as @steppi mentioned, this does not look like an appropriate error code for this situation.